kernel-rs/src/vm.rs

```diff
@@ -21,18 +21,454 @@ extern "C" {
+    type Target = [PageTableEntry; PTSIZE];
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl DerefMut for RawPageTable {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.inner
+    }
+}
+
+impl RawPageTable {
+    /// Look up a virtual address, return the physical address,
+    /// or 0 if not mapped.
+    /// Can only be used to look up user pages.
+    pub unsafe fn walkaddr(&mut self, va: usize) -> usize {
+        if va >= MAXVA {
+            return 0;
+        }
+        let pte_op = walk(self, va, 0);
+        if pte_op.is_none() {
+            return 0;
+        }
+        let pte = pte_op.unwrap();
+        if !pte.check_flag(PTE_V) {
+            return 0;
+        }
+        if !pte.check_flag(PTE_U as usize) {
+            return 0;
+        }
+        pte.get_pa()
+    }
+
+    /// Create PTEs for virtual addresses starting at va that refer to
+    /// physical addresses starting at pa. va and size might not
+    /// be page-aligned. Returns true on success, false if walk() couldn't
+    /// allocate a needed page-table page.
+    pub unsafe fn mappages(&mut self, va: usize, size: usize, mut pa: usize, perm: i32) -> bool {
+        let mut a = pgrounddown(va);
+        let last = pgrounddown(va.wrapping_add(size).wrapping_sub(1usize));
+        loop {
+            let pte = some_or!(walk(self, a, 1), return false);
+            if pte.check_flag(PTE_V) {
+                panic!("remap");
+            }
+            pte.set_inner(pa2pte(pa) | perm as usize | PTE_V);
+            if a == last {
+                break;
+            }
+            a = a.wrapping_add(PGSIZE);
+            pa = pa.wrapping_add(PGSIZE);
+        }
+        true
+    }
+
+    /// Recursively free page-table pages.
+    /// All leaf mappings must already have been removed.
+    unsafe fn freewalk(&mut self) {
+        // There are 2^9 = 512 PTEs in a page table.
+        for i in 0..512 {
+            let pte = &mut self[i];
+            if pte.check_flag(PTE_V) && !pte.check_flag((PTE_R | PTE_W | PTE_X) as usize) {
+                // This PTE points to a lower-level page table.
+                pte.as_table_mut().freewalk();
+                pte.set_inner(0);
+            } else if pte.check_flag(PTE_V) {
+                panic!("freewalk: leaf");
+            }
+        }
+        kfree(self.as_mut_ptr() as *mut libc::CVoid);
+    }
+
+    /// Remove mappings from a page table. The mappings in
+    /// the given range must exist. Optionally free the
+    /// physical memory.
+    pub unsafe fn uvmunmap(&mut self, va: usize, size: usize, do_free: i32) {
+        let mut pa: usize = 0;
+        let mut a = pgrounddown(va);
+        let last = pgrounddown(va.wrapping_add(size).wrapping_sub(1usize));
+        loop {
+            let pte_op = walk(self, a, 0);
+            if pte_op.is_none() {
+                panic!("uvmunmap: walk");
+            }
+            let pte = pte_op.unwrap();
```

Comment:
```suggestion
            let pte = walk(self, a, 0).expect("uvmunmap: walk");
```
일반적으로 `unwrap` 은 피하는 것이 좋습니다.